sdk/push-fcm/src/main/java/com/klaviyo/pushFcm/Extensions.kt

```diff
@@ -0,0 +1,213 @@
+import androidx.core.app.NotificationManagerCompat
+import androidx.core.content.PackageManagerCompat
+import androidx.core.content.res.ResourcesCompat
+import com.google.firebase.messaging.CommonNotificationBuilder
+import com.google.firebase.messaging.RemoteMessage
+import com.klaviyo.core.Registry
+
+/**
+ * Parse channel ID or fallback on default
+ */
+val RemoteMessage.channel_id: String
+    get() = this.data[Notification.CHANNEL_ID_KEY] ?: "Default"
+
+/**
+ * Parse channel name or fallback on default
+ */
+val RemoteMessage.channel_name: String
+    get() = this.data[Notification.CHANNEL_NAME_KEY] ?: "Default"
+
+/**
+ * Parse channel description or fallback on default
+ */
+val RemoteMessage.channel_description: String
+    get() = this.data[Notification.CHANNEL_DESCRIPTION_KEY]
+        ?: "Push notifications default channel"
+
+/**
+ * Parse channel importance or fallback on default
+ */
+val RemoteMessage.channel_importance: Int
+    get() = this.data[Notification.CHANNEL_IMPORTANCE_KEY]?.toInt()
+        ?: NotificationManagerCompat.IMPORTANCE_DEFAULT
+
+/**
+ * Parse out notification priority or fallback on default
+ */
+val RemoteMessage.notificationPriority: Int
+    get() = this.data[Notification.NOTIFICATION_PRIORITY]?.toInt()
+        ?: NotificationCompat.PRIORITY_DEFAULT
+
+/**
+ * Determine if the message originated from Klaviyo from the tracking params
+ */
+val RemoteMessage.isKlaviyoMessage: Boolean get() = this.data.containsKey("_k")
+
+/**
+ * Determine if the message is a notification from Klaviyo (as opposed to a silent push)
+ */
+val RemoteMessage.isKlaviyoNotification: Boolean
+    get() = this.isKlaviyoMessage && (title?.let { true } ?: body?.let { true } ?: false)
+
+/**
+ * Parse notification title text
+ */
+val RemoteMessage.title: String? get() = this.data[Notification.TITLE_KEY]
+
+/**
+ * Parse notification body text
+ */
+val RemoteMessage.body: String? get() = this.data[Notification.BODY_KEY]
+
+/**
+ * Parse deep link into a [Uri] if present
+ */
+val RemoteMessage.deepLink: Uri? get() = this.data[Notification.URL_KEY]?.let { Uri.parse(it) }
+
+/**
+ * Parse click action (activity or intent filter)
+ * Click action could be explicitly sent, or we should use ACTION_VIEW if a deep link is sent
+ */
+val RemoteMessage.clickAction: String?
+    get() = this.data[Notification.CLICK_ACTION_KEY] ?: deepLink?.let { Intent.ACTION_VIEW }
+
+/**
+ * Parse [Uri] to sound resource
+ */
+val RemoteMessage.sound: Uri? get() = this.data[Notification.SOUND_KEY]?.let { Uri.parse(it) }
+
+/**
+ * Parse out notification count from payload (for app badging)
+ */
+val RemoteMessage.notificationCount: Int
+    get() = this.data[Notification.NOTIFICATION_COUNT_KEY]?.toInt() ?: 1
+
+/**
+ * Determine the resource ID of the small icon
+ *
+ * NOTE: We have to use a discouraged API because we can't expect
+ *  developers to know the Int value of their icon resources
+ *
+ * @return
+ */
+val RemoteMessage.smallIcon: Int
```

Comment:
Get notification icon resource ID, first check the message payload, then the manifests, then fall back on system default.

Reply:
I like that this future proofs the SDK but are we adding too much code here that we may not get to test end to end sending stuff to the SDK?

Reply:
I wasn't too concerned about it since this is copied from how fcm handles icons

Reply:
I could run a test with my local just to be extra sure

Reply:
Is this icon stuff ripped from FCM?

Reply:
Oh actually I assume it is cause your comment on `isValidIcon` below

Reply:
And if yes, I guess the follow up question is - are there legal implications to copying that code? Like whats the license on FCM stuff

Reply:
Good question... I mean it isn't copy pasted but certainly influenced. I'll see what the license is and go from there

Reply:
Oh ok, sorry I assumed it was a copy/paste from your comment. I am probably over-thinking it then

Reply:
It's a valid concern for sure. I couldn't copy paste anyway since fcm is written in java. I feel like I had to make substantial enough changes to make it work for me that it is pretty distinct, even though I also feel comfortable saying we're using the same approach as they do so it should work the same.

Reply:
FWIW I hardcoded some FCM message payloads to send from my local, and confirmed the `small_icon` data key will work with this logic

Reply:
And firebase sdk is licensed under Apache License 2.0 so quite open.

Reply:
Thanks for testing this!